statsmodels/tsa/_stl.pyx

```diff
@@ -0,0 +1,680 @@
+    uses a correct partitioned sort to determine the median.
+
+    Examples
+    --------
+    The original example uses STL to decompos CO2 data into level, season and a
+    residual.
+
+    Start by aggregating to monthly, and filling any missing values
+
+    >>> from statsmodels.datasets import co2
+    >>> import matplotlib.pyplot as plt
+    >>> from pandas.plotting import register_matplotlib_converters
+    >>> register_matplotlib_converters()
+    >>> data = co2.load(True).data
+    >>> data = data.resample('M').mean().ffill()
+
+    The period (12) is automatically detected from the data's frequency ('M').
+
+    >>> from statsmodels.tsa.seasonal import STL
+    >>> res = STL(data).fit()
+    >>> res.plot()
+
+    See Also
+    --------
+    statsmodels.tsa.seasonal.STLResults
+    statsmodels.tsa.seasonal.seasonal_decompose
+
+    References
+    ----------
+    .. [1] R. B. Cleveland, W. S. Cleveland, J.E. McRae, and I. Terpenning
+        (1990) STL: A Seasonal-Trend Decomposition Procedure Based on LOESS.
+        Journal of Official Statistics, 6, 3-73.
+    """
+    cdef object endog
+    cdef Py_ssize_t nobs
+    cdef int period, seasonal, trend, low_pass, seasonal_deg, trend_deg
+    cdef int low_pass_deg, low_pass_jump, trend_jump, seasonal_jump
+    cdef bint robust, _use_rw
+    cdef double[::1] _ya, _trend, _season, _rw
+    cdef double[:, ::1] _work
+
+    def __init__(self, endog, period=None, seasonal=7, trend=None, low_pass=None,
+                 seasonal_deg=1, trend_deg=1, low_pass_deg=1,
+                 robust=False, seasonal_jump=1, trend_jump=1, low_pass_jump=1):
+        self.endog = endog
+        y = np.ascontiguousarray(np.squeeze(np.asarray(endog)), dtype=np.double)
+        if y.ndim != 1:
+            raise ValueError('y must be a 1d array')
+        self._ya = y
+        self.nobs = y.shape[0]  # n
+        if period is None:
+            freq = None
+            if isinstance(endog, (pd.Series, pd.DataFrame)):
+                freq = getattr(endog.index, 'inferred_freq', None)
+            if freq is None:
+                raise ValueError('Unable to determine period from endog')
+            period = freq_to_period(freq)
+        if not _is_pos_int(period, False) or period < 2:
+            raise ValueError('period must be a positive integer >= 2')
+        self.period = period  # np
+        if not _is_pos_int(seasonal, True) or seasonal < 3:
+            raise ValueError('seasonal must be an odd positive integer >= 3')
+        self.seasonal = seasonal  # ns
+        if trend is None:
+            trend = int(np.ceil(1.5 * self.period))
+            # ensure odd
+            trend += ((trend % 2) == 0)
+        if not _is_pos_int(trend, True) or trend < 3 or trend <= period:
+            raise ValueError('trend must be an odd positive integer '
+                             '>= 3 where trend > period')
+        self.trend = trend  # nt
```

Comment:
can you expand on this "nt", and "ns", "np" above?  I'm guessing these are for defaults of no-period, no-seasonal, and no-freq?

Reply:
They are defined in the header comment.  This is a port from FORTRAN and keeping this notation makes it simpler in case something goes wrong.